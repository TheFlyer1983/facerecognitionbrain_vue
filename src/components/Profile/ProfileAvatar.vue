<script setup lang="ts">
import useAvatars from '@/composables/avatars/useAvatars';
import { onMounted } from 'vue';

const props = defineProps<{
  name: string;
}>();

const { avatar, createAvatar } = useAvatars();

onMounted(() => {
  createAvatar(props.name);
});
</script>

<template>
  <div class="avatar d-flex justify-content-center align-items-center">
    {{ avatar }}
  </div>
</template>

<style lang="scss" scoped>
.avatar {
  border-radius: 100%;
  border: 1px solid black;
  display: inline-block;
  height: 4rem;
  width: 4rem;
  font-size: 2.25rem;
  font-weight: bold;
  background-color: #33b2f6;
}
</style>
