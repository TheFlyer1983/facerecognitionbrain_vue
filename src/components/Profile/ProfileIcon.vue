<script setup lang="ts">
import { ref } from 'vue';

const isDropdownOpen = ref(false);

function toggleDropdown() {
  isDropdownOpen.value = !isDropdownOpen.value;
}
</script>

<template>
  <section class="dropdown">
    <span @click="toggleDropdown" class="dropdown-button">
      <img
        src="http://tachyons.io/img/logo.jpg"
        alt="avatar"
        class="dropdown-button-image"
      />
    </span>
    <ul v-if="isDropdownOpen" class="dropdown-menu">
      <li class="dropdown-menu-item">View Profile</li>
      <li class="dropdown-menu-item">Sign Out</li>
    </ul>
  </section>
</template>

<style scoped lang="scss">
.dropdown {
  position: relative;
  margin: 1rem;

  &-button {
    display: flex;
    justify-content: flex-end;
    border: none;
    font-size: inherit;

    &-image {
      border-radius: 100%;
      border: 1px solid black;
      display: inline-block;
      height: 4rem;
      width: 4rem;
    }
  }

  &-menu {
    position: absolute;
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    right: 0;
    border: transparent;
    margin-top: 20px;
    background-color: rgba(255, 255, 255, 0.5);
    width: 10rem;
    list-style: none;
    padding-left: auto;
    padding-right: auto;
    padding-top: 1rem;
    border-radius: 5px;

    &-item {
      cursor: pointer;
      padding-bottom: 1rem;
    }
  }
}
</style>
