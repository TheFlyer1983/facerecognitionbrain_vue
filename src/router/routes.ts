export enum Routes {
  Home = 'home'
}