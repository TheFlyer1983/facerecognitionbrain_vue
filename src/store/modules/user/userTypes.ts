import { ActionContext } from 'vuex';

export type UserActionContext = ActionContext<UserState, {}>

export interface UserState {
  isSignedIn: boolean;
  isProfileOpen: false;
  user: User | null;
  token: string;
  rank: string;
}

export interface User {
  id: string;
  name: string;
  email: string;
  entries: number;
  joined: string;
  pet: string;
  age: string;
}
