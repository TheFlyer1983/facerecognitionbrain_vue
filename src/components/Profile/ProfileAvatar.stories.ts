import { ArgTypes, Story } from '@storybook/vue3';

import ProfileAvatar from './ProfileAvatar.vue';

export default {
  title: 'ProfileAvatar',
  component: ProfileAvatar
};

const Template: Story = (args) => ({
  components: { ProfileAvatar },
  setup() {
    return { args };
  },
  template: `<ProfileAvatar :name="args.name"/>`
});

export const Default: Story = Template.bind({});
Default.args = {
  name: 'Paul'
};
Default.argTypes = {
  name: {
    description: 'The logged in user name'
  }
} as ArgTypes;
