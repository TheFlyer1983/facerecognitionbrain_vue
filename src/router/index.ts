import { createRouter, createWebHistory, RouteRecordRaw } from 'vue-router';

import HelloWorld from '@/views/HelloWorld.vue';
import { Routes } from './routes';

export const INITIAL_ROUTE = '/';

const routes: Array<RouteRecordRaw> = [
  {
    path: INITIAL_ROUTE,
    name: 'Index',
    redirect: () => ({ name: Routes.Home })
  },
  {
    path: '/home',
    name: Routes.Home,
    component: HelloWorld,
    props: { msg: 'Hello Vue 3 + TypeScript + Vite' }
  },
  {
    path: '/login',
    name: Routes.Login,
    component: () => import('@/views/Login/Login.vue')
  },
  {
    path: '/register',
    name: Routes.Register,
    component: () => import('@/views/Register/Register.vue')
  }
];

const router = createRouter({
  history: createWebHistory(INITIAL_ROUTE),
  routes
});

export default router;
