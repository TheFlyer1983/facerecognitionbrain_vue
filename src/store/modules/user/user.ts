import { UserState, UserActionContext } from './userTypes';

import { endpoints } from '@/constants';
import { LoginInfo, RegisterInfo } from '@/types';
import {
  getAuthTokenInSession,
  saveAuthTokenInSession
} from '@/functions/storageFunctions';

export const state: UserState = {
  isSignedIn: false,
  isProfileOpen: false,
  user: null,
  token: ''
};

export const getters = {
  getUser(state: UserState): UserState['user'] {
    return state.user;
  },
  getToken(state: UserState): UserState['token'] {
    return state.token;
  },
  getIsSignedIn(state: UserState): UserState['isSignedIn'] {
    return state.isSignedIn;
  }
};

export const mutations = {
  setUserId(state: UserState, payload: string): void {
    state.user = { ...state.user, id: payload } as UserState['user'];
  },

  setUser(state: UserState, payload: UserState['user']): void {
    state.user = payload;
  },

  toggleSignIn(state: UserState, payload: UserState['isSignedIn']): void {
    state.isSignedIn = payload;
  },

  setToken(state: UserState, payload: UserState['token']): void {
    state.token = payload;
  }
};

export const actions = {
  async login(
    { commit, dispatch }: UserActionContext,
    payload: LoginInfo
  ): Promise<boolean> {
    try {
      const response = await fetch(endpoints.signin, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      const result = await response.json();

      if (!response.ok) throw new Error(result as string);

      saveAuthTokenInSession(result.token as string);
      commit('setToken', result.token);
      commit('setUserId', result.userId);

      const success = await dispatch('getUser', result.userId as string);
      if (!success) throw new Error('Error');

      return true;
    } catch (error) {
      console.log(error, 'In the error block');
      return false;
    }
  },
  async getUser({ commit }: UserActionContext, payload: string) {
    const requestURL = endpoints.getProfile.replace(':id', payload);
    const token = getAuthTokenInSession();
    try {
      if (!token) throw new Error('No valid token');
      const response = await fetch(requestURL, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          Authorization: token
        }
      });
      const result = await response.json();
      if (!response.ok) throw new Error(result);

      commit('setUser', result);
      commit('toggleSignIn', true);
      return true;
    } catch (error) {
      console.log(error);
      return false;
    }
  },
  async registerUser(
    { commit }: UserActionContext,
    payload: RegisterInfo
  ): Promise<boolean> {
    try {
      const response = await fetch(endpoints.register, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      const result = await response.json();

      if (!response.ok) throw new Error(result as string);

      saveAuthTokenInSession(result.session.token as string);
      commit('setUser', result.register.user);

      return true;
    } catch (error) {
      console.log(error);
      return false;
    }
  },
  getToken({ commit, dispatch }: UserActionContext) {
    const token = getAuthTokenInSession();

    if (token) {
      commit('setToken', token);
      dispatch('authenticated');
    }
  },
  async authenticated({ state, dispatch }: UserActionContext) {
    const token = state.token;

    if (token) {
      try {
        const response = await fetch(endpoints.signin, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Authorization: token
          }
        });

        const result = await response.json();

        if (!response.ok) throw new Error(result as string);

        dispatch('getUser', result.id);
      } catch (error) {
        console.log(error);
      }
    }
  }
};

const userModule = {
  namespaced: true,
  state,
  getters,
  mutations,
  actions
};

export default userModule;
