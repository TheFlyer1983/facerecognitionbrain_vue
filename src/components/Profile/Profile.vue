<script setup lang="ts">
import { ref, computed } from 'vue';
import { useUserStore } from '@/store/modules/user';

import Modal from '@/components/Modal/Modal.vue';
import { UpdateInfo } from '@/store/modules/user/userTypes';
import isEmpty from 'lodash.isempty';
import { useNavigation } from '@/modules/navigation';
import { Routes } from '@/router/routes';

const userStore = useUserStore();
const { navigate } = useNavigation();

function closeModal() {
  userStore.isProfileOpen = false;
}

const user = computed(() => userStore.user);

const userName = ref(null);
const age = ref<number | null>(null);
const agePlaceHolderText = computed(() => {
  if (user.value?.age) return user.value.age.toString();
  return '';
});
const pet = ref(null);

async function updateProfile() {
  let userInfo: UpdateInfo = {};

  if (userName.value && userName.value !== user.value?.name)
    userInfo = { ...userInfo, name: userName.value };
  if (age.value && age.value !== user.value?.age)
    userInfo = { ...userInfo, age: age.value };
  if (pet.value && pet.value !== user.value?.pet)
    userInfo = { ...userInfo, pet: pet.value };

  if (!isEmpty(userInfo)) await userStore.updateUser(userInfo);
  closeModal();
}

const showConfirmation = ref(false);

function toggleConfirmation() {
  showConfirmation.value = !showConfirmation.value;
}

async function deleteUser() {
  await userStore.deleteUser();
  navigate({ name: Routes.Login });
}
</script>

<template>
  <Modal>
    <div class="profile-modal">
      <article
        class="br3 ba b--black-10 mv4 w-100 w-50-m w-25-l mw6 shadow-5 bg-white d-flex justify-content-center"
      >
        <main class="pa4 black-80 w-80">
          <img
            src="http://tachyons.io/img/logo.jpg"
            alt="avatar"
            class="h3 w3 dib"
          />
          <h1>{{ user?.name }}</h1>
          <h4>Images Submitted: {{ user?.entries }}</h4>
          <p>
            Member Since:
            {{ new Date(`${user?.joined}`).toLocaleDateString('en-GB') }}
          </p>
          <hr />
          <template v-if="showConfirmation">
            <p>Are you sure you want to delete your profile?</p>
            <p>This is cannot be undone!</p>
            <div class="button-group mt4">
              <button
                class="b pa2 grow pointer hover-white w-40 bg-light-blue b--black-20"
                @click="deleteUser"
                data-test="confirm"
              >
                Yes
              </button>
              <button
                class="b pa2 grow pointer hover-white w-40 bg-light-red b--black-20"
                @click="closeModal"
                data-test="cancel-delete"
              >
                No
              </button>
            </div>
          </template>
          <template v-else>
            <label for="user-name" class="mv2 fw6">Name:</label>
            <input
              type="text"
              name="user-name"
              id="name"
              class="pa2 ba w-100"
              v-model="userName"
              :placeholder="user?.name"
            />
            <label for="user-age" class="mv2 fw6">Age:</label>
            <input
              type="text"
              name="user-age"
              id="age"
              class="pa2 ba w-100"
              v-model="age"
              :placeholder="agePlaceHolderText"
            />
            <label for="user-pet" class="mv2 fw6">Pet:</label>
            <input
              type="text"
              name="user-pet"
              id="pet"
              class="pa2 ba w-100"
              v-model="pet"
              :placeholder="user?.pet"
            />
            <div class="button-group mt4">
              <button
                class="b pa2 grow pointer hover-white w-40 bg-light-blue b--black-20"
                @click="updateProfile"
              >
                Save
              </button>
              <button
                class="b pa2 grow pointer hover-white w-40 bg-light-red b--black-20"
                @click="closeModal"
              >
                Cancel
              </button>
            </div>
            <hr />
            <div class="button-group mt4">
              <button
                class="b pa2 grow pointer hover-white w-40 bg-light-red b--black-20"
                @click="toggleConfirmation"
                data-test="delete"
              >
                Delete User
              </button>
            </div>
          </template>
        </main>
        <div class="modal-close" @click="closeModal">&times;</div>
      </article>
    </div>
  </Modal>
</template>

<style lang="scss" scoped>
.button-group {
  display: flex;
  justify-content: space-evenly;
}
.modal-close {
  font-size: 3rem;
  font-style: bold;
  cursor: pointer;

  &:hover {
    color: grey;
  }
}
</style>
