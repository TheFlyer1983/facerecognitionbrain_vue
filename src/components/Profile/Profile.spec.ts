import { shallowMount, VueWrapper } from '@vue/test-utils';
import { reactive } from 'vue';
import { useStore } from 'vuex';
import { useNavigation } from '@/modules/navigation';

import Profile from './Profile.vue';

import { UserMock } from '@/fixtures/users';
import { User } from '@/store/modules/user/userTypes';
import { Routes } from '@/router/routes';

vi.mock('vuex');
vi.mock('@/modules/navigation');

const navigateMock = vi.fn();

const mockedUseStore = vi.mocked<() => Partial<typeof useStore>>(useStore);
const mockedUseNavigation = vi.mocked(useNavigation, true);

const mockedGetters = {
  ['user/getUser']: UserMock as User | null
};

const mockedStore = reactive({
  getters: mockedGetters,
  commit: vi.fn(),
  dispatch: vi.fn()
});

describe('Given the `Profile` component', () => {
  const render = () =>
    shallowMount(Profile, {
      global: {
        stubs: {
          Modal: false
        }
      }
    });
  let wrapper: VueWrapper<InstanceType<typeof Profile>>;

  beforeAll(() => {
    mockedUseStore.mockReturnValue(mockedStore);
    mockedUseNavigation.mockReturnValue({ navigate: navigateMock });
  });

  describe('and when the component is rendered', () => {
    beforeEach(() => {
      wrapper = render();
    });

    it('should match the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('when the close modal function is called', () => {
      beforeEach(() => {
        wrapper.vm.closeModal();
      });

      it('should commit the correct mutation', () => {
        expect(mockedStore.commit).toHaveBeenCalledWith('user/toggleModal');
      });
    });

    describe('and when the user getter returns no age value', () => {
      beforeEach(() => {
        mockedStore.getters['user/getUser'] = null;
      });

      it('and the placeholder text should be blank', () => {
        expect(wrapper.vm.agePlaceHolderText).toBe('');
      });
    });

    describe('and when the `updateProfile` function is called', () => {
      beforeEach(() => {
        mockedStore.getters['user/getUser'] = UserMock;
      });

      describe('and when there are no values entered', () => {
        beforeEach(() => {
          wrapper.vm.updateProfile();
        });

        it('should not dispatch the action', () => {
          expect(mockedStore.dispatch).not.toHaveBeenCalled();
        });
      });

      describe.each([
        { key: UserMock.name, inputField: 'userName' },
        { key: UserMock.age, inputField: 'age' },
        { key: UserMock.pet, inputField: 'pet' }
      ])(
        'and when the userName has been entered, but has the same value',
        (arg) => {
          beforeEach(() => {
            wrapper.vm[arg.inputField] = arg.key;
            wrapper.vm.updateProfile();
          });

          it('should not dispatch the action', () => {
            expect(mockedStore.dispatch).not.toHaveBeenCalled();
          });
        }
      );

      describe.each([
        {
          key: 'Peter',
          inputField: 'userName',
          property: 'name'
        },
        {
          key: 40,
          inputField: 'age',
          property: 'age'
        },
        {
          key: 'Audrey',
          inputField: 'pet',
          property: 'pet'
        }
      ])('and when the userName has been entered', (arg) => {
        const userInfo = {
          [arg.property]: arg.key
        };
        beforeEach(() => {
          wrapper.vm[arg.inputField] = arg.key;
          wrapper.vm.updateProfile();
        });

        it('should not dispatch the action', () => {
          expect(mockedStore.dispatch).toHaveBeenCalledWith(
            'user/updateUser',
            userInfo
          );
        });
      });
    });

    describe('and when then `Delete` button is clicked', () => {
      beforeEach(() => {
        wrapper.find('[data-test="delete"]').trigger('click');
      });

      it('should update the `showConfirmation` value', () => {
        expect(wrapper.vm.showConfirmation).toBe(true);
      });

      it('should match the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      describe('and when the `Yes` button is clicked', () => {
        beforeEach(() => {
          wrapper.find('[data-test="confirm"]').trigger('click');
        });

        it('should dispatch the correct actions', () => {
          expect(mockedStore.dispatch).toHaveBeenCalledWith('user/deleteUser');

          expect(mockedStore.dispatch).toHaveBeenCalledWith('user/signOut');
        });

        it('should navigate to the correct page', () => {
          expect(navigateMock).toHaveBeenCalledWith({ name: Routes.Login });
        });
      });

      describe('and when the `Cancel` button is clicked', () => {
        beforeEach(() => {
          wrapper.find('[data-test="cancel-delete"]').trigger('click');
        });

        it('should commit the correct mutation', () => {
          expect(mockedStore.commit).toHaveBeenCalledWith('user/toggleModal');
        });
      });
    });
  });
});
