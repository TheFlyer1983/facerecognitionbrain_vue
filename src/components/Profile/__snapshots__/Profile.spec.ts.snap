// Vitest Snapshot v1

exports[`Given the \`Profile\` component > and when the component is rendered > and when then \`Delete\` button is clicked > should match the snapshot 1`] = `
<div
  class="modal position-abosolute d-flex justify-content-center align-items-center"
  data-v-2712ae82=""
  data-v-907877e9=""
>
  <div
    class="modal__overlay position-fixed w-100 h-100 bg-white-80 top-0 start-0"
    data-test="close-modal"
    data-v-907877e9=""
  />
  <div
    class="modal__container"
    data-v-907877e9=""
  >
    
    <div
      class="profile-modal"
      data-v-2712ae82=""
    >
      <article
        class="br3 ba b--black-10 mv4 w-100 w-50-m w-25-l mw6 shadow-5 bg-white d-flex justify-content-center"
        data-v-2712ae82=""
      >
        <main
          class="pa4 black-80 w-80"
          data-v-2712ae82=""
        >
          <img
            alt="avatar"
            class="h3 w3 dib"
            data-v-2712ae82=""
            src="http://tachyons.io/img/logo.jpg"
          />
          <h1
            data-v-2712ae82=""
          >
            Paul
          </h1>
          <h4
            data-v-2712ae82=""
          >
            Images Submitted: 1
          </h4>
          <p
            data-v-2712ae82=""
          >
             Member Since: 09/04/2021
          </p>
          <hr
            data-v-2712ae82=""
          />
          
          <p
            data-v-2712ae82=""
          >
            Are you sure you want to delete your profile?
          </p>
          <p
            data-v-2712ae82=""
          >
            This is cannot be undone!
          </p>
          <div
            class="button-group mt4"
            data-v-2712ae82=""
          >
            <button
              class="b pa2 grow pointer hover-white w-40 bg-light-blue b--black-20"
              data-test="confirm"
              data-v-2712ae82=""
            >
               Yes 
            </button>
            <button
              class="b pa2 grow pointer hover-white w-40 bg-light-red b--black-20"
              data-test="cancel-delete"
              data-v-2712ae82=""
            >
               No 
            </button>
          </div>
          
        </main>
        <div
          class="modal-close"
          data-v-2712ae82=""
        >
          ×
        </div>
      </article>
    </div>
    
  </div>
</div>
`;

exports[`Given the \`Profile\` component > and when the component is rendered > should match the snapshot 1`] = `
<div
  class="modal position-abosolute d-flex justify-content-center align-items-center"
  data-v-2712ae82=""
  data-v-907877e9=""
>
  <div
    class="modal__overlay position-fixed w-100 h-100 bg-white-80 top-0 start-0"
    data-test="close-modal"
    data-v-907877e9=""
  />
  <div
    class="modal__container"
    data-v-907877e9=""
  >
    
    <div
      class="profile-modal"
      data-v-2712ae82=""
    >
      <article
        class="br3 ba b--black-10 mv4 w-100 w-50-m w-25-l mw6 shadow-5 bg-white d-flex justify-content-center"
        data-v-2712ae82=""
      >
        <main
          class="pa4 black-80 w-80"
          data-v-2712ae82=""
        >
          <img
            alt="avatar"
            class="h3 w3 dib"
            data-v-2712ae82=""
            src="http://tachyons.io/img/logo.jpg"
          />
          <h1
            data-v-2712ae82=""
          >
            Paul
          </h1>
          <h4
            data-v-2712ae82=""
          >
            Images Submitted: 1
          </h4>
          <p
            data-v-2712ae82=""
          >
             Member Since: 09/04/2021
          </p>
          <hr
            data-v-2712ae82=""
          />
          
          <label
            class="mv2 fw6"
            data-v-2712ae82=""
            for="user-name"
          >
            Name:
          </label>
          <input
            class="pa2 ba w-100"
            data-v-2712ae82=""
            id="name"
            name="user-name"
            placeholder="Paul"
            type="text"
          />
          <label
            class="mv2 fw6"
            data-v-2712ae82=""
            for="user-age"
          >
            Age:
          </label>
          <input
            class="pa2 ba w-100"
            data-v-2712ae82=""
            id="age"
            name="user-age"
            placeholder="37"
            type="text"
          />
          <label
            class="mv2 fw6"
            data-v-2712ae82=""
            for="user-pet"
          >
            Pet:
          </label>
          <input
            class="pa2 ba w-100"
            data-v-2712ae82=""
            id="pet"
            name="user-pet"
            placeholder="Ruby"
            type="text"
          />
          <div
            class="button-group mt4"
            data-v-2712ae82=""
          >
            <button
              class="b pa2 grow pointer hover-white w-40 bg-light-blue b--black-20"
              data-v-2712ae82=""
            >
               Save 
            </button>
            <button
              class="b pa2 grow pointer hover-white w-40 bg-light-red b--black-20"
              data-v-2712ae82=""
            >
               Cancel 
            </button>
          </div>
          <hr
            data-v-2712ae82=""
          />
          <div
            class="button-group mt4"
            data-v-2712ae82=""
          >
            <button
              class="b pa2 grow pointer hover-white w-40 bg-light-red b--black-20"
              data-test="delete"
              data-v-2712ae82=""
            >
               Delete User 
            </button>
          </div>
          
        </main>
        <div
          class="modal-close"
          data-v-2712ae82=""
        >
          ×
        </div>
      </article>
    </div>
    
  </div>
</div>
`;
